### DISCUSSION TITLE
Should image automation CRD's live in flux-system or their own namespaces?

### DISCUSSION THREAD

**Discussion by ilya-git** [fluxcd/flux2#1017] [Q&A]
I am trying to add few repositories to be automated by flux, e.g.
- team1-gitops git repository will contain manifests for team1 namespace
- team2-gitops git repository will contain manifests for team2 namespace

The documentation now uses `flux-system` namespace for all it's examples for ImageAutomation. If I add 2 image automations to the the same namespace it seems that both automations will scan all image policies which is not ideal, but perhaps not a big deal since comment marks will be correct. While 2 namespaces will mean I'll have to create e.g. 2 secrets to connect to docker which will be essentially the same and probably some other configuration as well.

What I am trying to understand here: should I put all these CRD's in flux-system namespace or is it better to actually split them in their own `team1`, `team2` namespaces? It would be very nice to have some guidance on namespaces as this is a new feature for flux that was not available before and it's a bit hard to figure out what is the proper (or recommended) way of doing it.

**Comment by kingdonb** [comment]
First, to check my own understanding, I am pretty sure that `ImagePolicy` itself is not a thing which does the scanning, it's just a record of the policy decision that lives in a CR. The `image-automation-controller` does the work, with the help of `image-reflector-controller`, and ideally as you only have one instance of each on your cluster, you can opt to use as many setters here and there that point to the same policies from within different namespaces without concern about scaling problems.

As for where an `ImagePolicy` itself should live: I think the answer is implementation-dependent and Flux is intentionally non-prescriptive about this, since it depends on your specific environment. If team1 has 3 apps, and team2 has 3 completely different apps, one way that makes sense to me is to put `imagepolicy` of record for each app in the namespace with the app, and let the team manage the policies that fall within their namespace. This seems like a "devops" answer, since AIUI according to devops principles, teams should control their stack from end to end, and that would seem to include decisions about `ImagePolicy`. If there is overlap between teams, then the answer is each team could maintain their own separate policies, or a central policy could be used. 

Some organizations will instead lean in the direction that for anything described as a "policy" it should be centrally managed and carefully change-controlled, with a board of people who decide whether and when a change is appropriate or should be shelved, and a (say) weekly meeting to discuss and decide about these concerns as a group. Hopefully it is clearer now why this is not something for Flux to be prescriptive about (but perhaps could add more docs to be clearer about the way we might decide these issues.)

**Reply by ilya-git** [reply]
Moving ImagePolicy to an own namespace would mean ImageUpdateAutomation, GitRepository, etc. would have to be moved there since they all reference CRD's within their own namespace only, including e.g. a secret used to access docker repository. That would mean duplication of the same secret in various namespaces, that's why it's a bit more than just a name of a namespace that has to be changed. But perhaps you are right and flux team did that on purpose (not mentioned where they should live).

**Reply by kingdonb** [reply]
I was not implying that any omission from the doc is on purpose (we should resolve this with more docs, I think I meant that I assume if the information is not present, it just means we haven't got the light there yet.) I would guess that if the two teams are separate tenants then they would also have their own secret credential / bot user, so there is no duplication.

There should be another multi-tenancy example added I think with a bit more different detail in it. I think the example considers "dev" and "ops" to be the two tenants, with dev shipping changes to ops. There is one `dev-team` in the example and most people I think when looking for a multi-tenancy example are expecting to see more teams (eg in your example team1 shares resources with team2 in the form of a shared cluster, but they do not need to intersect otherwise... although in places I've worked, they always do intersect at the Ops team, which handles production resources for both teams.)

So the example is valid but IMHO still needs more elaboration for realism in this way.

For the users that I've talked this through with, I think I have come to the conclusion that each team with multiple namespaces will have a "home namespace" where their automations live to avoid this kind of duplication as you go further down the example tree. So in the sense that the tutorial shows `flux-system` as the home namespace for all of Flux's machinery, with tenant apps living in the `apps` namespace, ... this could get extremely complicated very quickly depending on the number of teams and apps, and degree of isolation. So we should take account of this in our (individual) multi-tenancy designs, I do believe you're thinking about this in the right way.

**Reply by ilya-git** [reply]
I thought that perhaps allowing ImageUpdateAutomation to reference another namespace would be very nice. All machinery can live in flux-system, while the ImagePolicy can live in it's onw namespace with automation referencing it. Would probably be the best of both worlds, but so far it's not possible I think

**Reply by Frizlab** [reply]
I think the same; it would very much like if ImageUpdateAutomation checked all namespaces instead of only its own (or at least had the possibility to do so).

**Comment by 181192** [comment]
As an "DevOps" engineer I want to bootstrap clusters with resources like monitoring, logging, ingress controller, flux etc... And then give access to the cluster for multiple development teams where each team could have multiple namespaces. The development teams have full responsibility for managing their own applications, including releasing the their applications (you build it, you run it).
When using Flux v1 we had one Flux instance that managed the platform components (one repository + kustomize to manage all clusters), and one or several other instances for the development teams (one Flux per repository).

We have now started on migrating to Flux v2, for the platform component this is easy. Just regular Flux v2 setup with read access to sync platform components, low or no need to have image update on these components because changing etc Ingress Controller version would require testing and verification in a test environment before applying to production. So these things have been done manually when it comes to bump chart versions or image versions.

But introducing Flux v2 to the development teams with `ImageUpdateAutomation`, `ImagePolicy` and `ImageRepository` has not been successful yet. I'm not finding a good solution where to store these manifests, and maybe its the `ImageUpdateAutomation` crd that is the root of the issue. `ImageUpdateAutomation` refrence a `GitRepository`, and in our setup the "DevOps" team would not know about the namespaces required, only the Git repository for the development team, so what I have done now is creating the `GitRepository` and `Kustomization` in `flux-system` namespace. Because this is the "only" information I've got at this moment, just a normal streamlined cattle cluster, should not need to know anything more. And now `ImageUpdateAutomation` would fail because it wont read the `GitRepository` from another namespace. If I did it the other way around it would mean duplicating secrets and `GitRepository/Kustomization` crds and require to have more information about namespaces the development teams needs (and creates on their own).

Flux v1 is simple and would just commit back to the repository, I maybe expected the same behavior due to the `{"$imagepolicy": "<policy-namespace>:<policy-name>"}` policy marker that the image controller would figure out what manifests was tracked by which `Kustomization` manifest and use the same credentials. And not needing the `ImageUpdateAutomation` crd...

**Reply by ilya-git** [reply]
Totally agree, same question I face with the namespaces here, perhaps referencing git repository from another namespace will fix it?

**Reply by Legion2** [reply]
I also have a namespace problem. I use kustomize to manage multiple applications in different namespaces. I organize the manifests into directories for each namespace and each directory contains a `kustomization.yaml`, which overrides the namespace for all manifests in that directory. All manifests are synced from the flux-system `GitRepository` and flux-system `Kustomization`. Now to enable image update automation, I have to add the `ImageUpdateAutomation` into the flux-system namespace and also the `ImagePolicy` and the `ImageRepository` have to be in the flux-system namespace.

Now I want to add the `ImagePolicy` and the `ImageRepository` into the directory/namespace where the images are used, but then kustomize overrides the namespace of them and so the `ImageUpdateAutomation` does not find them.

A workaround would be to place all `ImagePolicy` and the `ImageRepository` into a separate directory with the flux-system namespace, but I want to keep them together with the application manifests in the same directory.

**Comment by ilya-git** [comment]
There is a discussion going on in [here](https://github.com/fluxcd/image-automation-controller/issues/85) about possible solution
